app.js

```diff
@@ -52,88 +53,84 @@ module.exports = function(someCodeObjects) {
-    return {objectIds: objectIds};
-  }
-
-  function broadcastObjectList(playgroundId) {
-    io.to(playgroundId).emit('objects list', getListOfAllObjects(playgroundId));
-  }
 
   io.on('connection', function(socket) {
     var query = socket.handshake.query;
-    var playgroundId = query.playgroundId;
+    var playground = world.getOrCreatePlayground(query.playgroundId);
     var client =  query.client;
 
-    socket.join(playgroundId);
-
     if (client == 'renderer')
       rendererUp();
     else
       programmerUp();
 
     function programmerUp() {
-      debug("a new programmer is up for " + playgroundId);
+      debug("a new programmer is up for " + playground.id);
 
-      if (!codeObjects[playgroundId]) return;
-      socket.emit('objects list', getListOfAllObjects(playgroundId));
+      if (!playground.isEmpty())
+        sendListOfAllOBjects();
     }
 
+
     function rendererUp() {
-      debug("a new renderer is up for " + playgroundId);
+      debug("a new renderer is up for " + playground.id);
 
-      if (!codeObjects[playgroundId]) return;
-      socket.emit('playground full update', codeObjects[playgroundId]);
-    }
+      if (playground.isEmpty()) return;
 
-    socket.on('code update', function(data) {
-      var objectId = data.objectId;
+      var data = Object.create(null);
+      playground.population().forEach(function(codeObjectId) {
+        data[codeObjectId] = playground.getOrCreateCodeObject(codeObjectId).getData();
+      });
+      socket.emit('playground full update', data);
+    }
 
-      debug(objectId + " for " + playgroundId + " from " + client);
+    function sendListOfAllOBjects() {
+      socket.emit('objects list', {objectIds: playground.population()});
+    }
 
-      if (!codeObjects[playgroundId]) codeObjects[playgroundId] = {};
+    socket.on('code update', function(data) {
+      debug(data.objectId + " for " + playground.id + " from " + data.client);
 
-      codeObjects[playgroundId][objectId] = {
-        mediatype: data.mediatype,
-        client: client,
-        code: data.code
-      };
+      var codeObject = playground.getOrCreateCodeObject(data.objectId);
 
-      socket.broadcast.to(playgroundId).emit('code update', data);
-      broadcastObjectList(playgroundId);
+      codeObject.setData(data);
     });
 
     socket.on('code delete', function(data) {
-      var objectId = data.objectId;
+      debug("deleting " + data.objectId + " from playground " + playground.id);
 
-      if (!codeObjects[playgroundId]) return;
+      if (playground.contains(data.objectId))
+          playground.deleteCodeObject(data.objectId);
+    });
 
-      debug("deleting " + objectId + " from playground " + playgroundId);
+    socket.on('request code', function(data) {
+      debug(data.objectId + " for " + playground.id + " programmer" ) ;
 
-      delete codeObjects[playgroundId][objectId];
+      if (!playground.contains(data.objectId)) return;
```

Comment:
This `return` could be the source of the [regression](https://github.com/jonathanperret/paysage/pull/62#issuecomment-325329587) I hit.

The server used to always reply to `request code` with a `source code` message (with empty code if necessary).